scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```diff
@@ -0,0 +1,159 @@
+  }
+
+  /**
+   * Traverse given schemas and format node for list or map of projected read type to structure
+   * of file schema. The formatting is not to one-to-one node swapping between the two schemas
+   * because of the projection requirement.
+   */
+  private def projectFileType(projectedReadType: Type, fileType: Type, fieldContext: FieldContext): Type = {
+    (extractCollectionGroup(projectedReadType), extractCollectionGroup(fileType)) match {
+      case _ if projectedReadType.isPrimitive && fileType.isPrimitive =>
+        projectedReadType
+      case _ if projectedReadType.isPrimitive != fileType.isPrimitive =>
+        throw new DecodingSchemaMismatchException(
+          s"Found schema mismatch between projected read type:\n$projectedReadType\n" +
+            s"and file type:\n${fileType}"
+        )
+      case (Some(projectedReadGroup: ListGroup), Some(fileGroup: ListGroup)) =>
+        projectFileGroup[ListGroup](projectedReadGroup, fileGroup, fieldContext)
+      case (Some(projectedReadGroup: MapGroup), Some(fileGroup: MapGroup)) =>
+        projectFileGroup[MapGroup](projectedReadGroup, fileGroup, fieldContext)
+      case _ => // Field projection
+        val projectedReadGroupType = projectedReadType.asGroupType
+        val fileGroupType = fileType.asGroupType
+        val projectedReadFields = projectedReadGroupType.getFields.asScala.map { projectedReadField =>
+          if (!fileGroupType.containsField(projectedReadField.getName)) {
+            if (!projectedReadField.isRepetition(Repetition.OPTIONAL)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found non-optional projected read field ${projectedReadField.getName}:\n$projectedReadField\n\n" +
+                  s"not present in the given file group type:\n${fileGroupType}"
+              )
+            }
+            projectedReadField
+          }
+          else {
+            val fileFieldIndex = fileGroupType.getFieldIndex(projectedReadField.getName)
+            val fileField = fileGroupType.getFields.get(fileFieldIndex)
+            projectFileType(projectedReadField, fileField, FieldContext(projectedReadField.getName))
+          }
+        }
+        projectedReadGroupType.withNewFields(projectedReadFields.asJava)
+    }
+  }
+
+  private def projectFileGroup[T <: CollectionGroup](projectedReadGroup: T,
+                                                     fileGroup: T,
+                                                     fieldContext: FieldContext)(implicit t: ClassTag[T]): GroupType = {
+
+    val (formatter, updatedFieldContext) = t.runtimeClass.asInstanceOf[Class[T]] match {
+      case c if c == classOf[MapGroup] =>
+        (ParquetMapFormatter, fieldContext.copy(nestedListLevel = fieldContext.nestedListLevel + 1))
+      case c if c == classOf[ListGroup] =>
+        (ParquetListFormatter, fieldContext.copy(nestedListLevel = fieldContext.nestedListLevel + 1))
+    }
```

Comment:
Let's refactor this to avoid reflection and implicit magic. E.g.:

```suggestion
private def projectFileGroup(projectedReadGroup: T, fileGroup: T, fieldContext: FieldContext, formatter: ParquetCollectionFormatter): GroupType = {
```

Reply:
Updated to
```
def projectFileGroup(fileGroup: CollectionGroup,
                               projectedReadGroup: CollectionGroup,
                               fieldContext: FieldContext,
                               formatter: ParquetCollectionFormatter)
```